server/bin/pbench-tarball-stats

```diff
@@ -0,0 +1,193 @@
+                        # flag this unwanted condition.
+                        continue
+                    match = tb_pat.fullmatch(entry.name)
+                    if match:
+                        # Turn the pattern components of the match into a datetime
+                        # object.
+                        tb_dt = datetime(
+                            int(match.group(1)),
+                            int(match.group(2)),
+                            int(match.group(3)),
+                            int(match.group(4)),
+                            int(match.group(5)),
+                            int(match.group(6)),
+                        )
+                    else:
+                        if entry.name.endswith(".md5"):
+                            continue
+                        tb_dt = None
+                    yield c_entry.name, entry.name, tb_dt, entry.stat()
+
+
+def main(options):
+    locale.setlocale(locale.LC_ALL, "en_US.UTF-8")
+    if not options.cfg_name:
+        print(
+            f"{_NAME_}: ERROR: No config file specified; set"
+            " _PBENCH_SERVER_CONFIG env variable",
+            file=sys.stderr,
+        )
+        return 1
+
+    try:
+        config = PbenchConfig(options.cfg_name)
+    except BadConfig as e:
+        print(f"{_NAME_}: {e}", file=sys.stderr)
+        return 2
+
+    logger = get_pbench_logger(_NAME_, config)
+
+    archive = config.ARCHIVE
+    archive_p = os.path.realpath(archive)
+
+    if not archive_p:
+        logger.error("The configured ARCHIVE directory, {}, does not exist", archive)
+        return 3
+
+    if not os.path.isdir(archive_p):
+        logger.error(
+            "The configured ARCHIVE directory, {}, is not a valid directory", archive
+        )
+        return 4
+
+    invalid_cnt = 0
+    invalid_size = 0
+    invalid = collections.defaultdict(list)
+
+    good_cnt = 0
+    good_size = 0
+    good = collections.defaultdict(list)
+
+    ec2_cnt = 0
+    rh_cnt = 0
+
+    by_year = collections.defaultdict(list)
+    by_month = collections.defaultdict(list)
+
+    start = pbench._time()
+
+    gen = gen_tar_balls(archive_p)
+
+    for tb_rec in gen:
+        if tb_rec[2] is None:
+            invalid[tb_rec[0]].append(tb_rec)
+            invalid_cnt += 1
+            invalid_size += tb_rec[3].st_size
+        else:
+            good[tb_rec[0]].append(tb_rec)
+            good_cnt += 1
+            good_size += tb_rec[3].st_size
+            if tb_rec[0].startswith("EC2::"):
+                ec2_cnt += 1
+            else:
+                rh_cnt += 1
+            year = tb_rec[2].year
+            by_year[year].append(tb_rec)
+            month = tb_rec[2].month
+            ym = f"{year:4d}-{month:02d}"
+            by_month[ym].append(tb_rec)
+
+    time_sec = pbench._time() - start
+
+    print(f"Took {time_sec:0.2f} seconds")
+
+    print(f"{len(invalid.keys()):18n} controllers with bad tar balls")
```

Comment:
How can `len(invalid.keys())` ever be different from `invalid_cnt`? If you're looking at both anyway, why not just use the `len()` SPOT?

Reply:
Count of controllers with invalid tar balls vs total count of invalid tar balls.